In [1]:
# data analysis and wrangling
import pandas as pd
import numpy as np
import random as rnd

# visualization
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

# machine learning
#from sklearn.linear_model import LogisticRegression
#from sklearn.svm import SVC, LinearSVC
#from sklearn.ensemble import RandomForestClassifier
#from sklearn.neighbors import KNeighborsClassifier
#from sklearn.naive_bayes import GaussianNB
#from sklearn.linear_model import Perceptron
#from sklearn.linear_model import SGDClassifier
#from sklearn.tree import DecisionTreeClassifier

In [2]:
data_train = pd.read_csv('train.csv') #Never place header = None
data_test =  pd.read_csv('test.csv')
data_train.columns = ['Pid','Survived', 'Pclass','Name','Sex','Age','Sibsp','Parch','Ticket','Fare','Cabin','Embarked']
data_test.columns =  ['Pid', 'Pclass','Name','Sex','Age','Sibsp','Parch','Ticket','Fare','Cabin','Embarked']

print(len(data_train));
print(len(data_test));


891
418


In [3]:
data_train.info()
data_test.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Pid       891 non-null    int64  
 1   Survived  891 non-null    int64  
 2   Pclass    891 non-null    int64  
 3   Name      891 non-null    object 
 4   Sex       891 non-null    object 
 5   Age       714 non-null    float64
 6   Sibsp     891 non-null    int64  
 7   Parch     891 non-null    int64  
 8   Ticket    891 non-null    object 
 9   Fare      891 non-null    float64
 10  Cabin     204 non-null    object 
 11  Embarked  889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Pid       418 non-null    int64  
 1   Pclass    418 non-null    int64  
 2   Name      418 non-nu

In [4]:
#data_train = data_train.iloc[0:]
data_train.drop(data_train.head(1).index, inplace=True)
print(data_train.head())
data_test.drop(data_test.head(1).index, inplace=True)
print(data_test.head())

   Pid  Survived  Pclass                                               Name  \
1    2         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2    3         1       3                             Heikkinen, Miss. Laina   
3    4         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4    5         0       3                           Allen, Mr. William Henry   
5    6         0       3                                   Moran, Mr. James   

      Sex   Age  Sibsp  Parch            Ticket     Fare Cabin Embarked  
1  female  38.0      1      0          PC 17599  71.2833   C85        C  
2  female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
3  female  35.0      1      0            113803  53.1000  C123        S  
4    male  35.0      0      0            373450   8.0500   NaN        S  
5    male   NaN      0      0            330877   8.4583   NaN        Q  
   Pid  Pclass                                          Name     Sex   Age  \
1  

In [5]:
data_train = data_train.drop(['Pid'],axis = 1)
data_test = data_test.drop(['Pid'],axis = 1)

In [6]:
data_train.head()

Survived  Pclass                                               Name  \
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2         1       3                             Heikkinen, Miss. Laina   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4         0       3                           Allen, Mr. William Henry   
5         0       3                                   Moran, Mr. James   

      Sex   Age  Sibsp  Parch            Ticket     Fare Cabin Embarked  
1  female  38.0      1      0          PC 17599  71.2833   C85        C  
2  female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
3  female  35.0      1      0            113803  53.1000  C123        S  
4    male  35.0      0      0            373450   8.0500   NaN        S  
5    male   NaN      0      0            330877   8.4583   NaN        Q

In [7]:
#data_train.to_csv("training-data.csv")

In [7]:
print(data_train.tail())
print("\n Columns are \n")
print(data_train.columns)

     Survived  Pclass                                      Name     Sex   Age  \
886         0       2                     Montvila, Rev. Juozas    male  27.0   
887         1       1              Graham, Miss. Margaret Edith  female  19.0   
888         0       3  Johnston, Miss. Catherine Helen "Carrie"  female   NaN   
889         1       1                     Behr, Mr. Karl Howell    male  26.0   
890         0       3                       Dooley, Mr. Patrick    male  32.0   

     Sibsp  Parch      Ticket   Fare Cabin Embarked  
886      0      0      211536  13.00   NaN        S  
887      0      0      112053  30.00   B42        S  
888      1      2  W./C. 6607  23.45   NaN        S  
889      0      0      111369  30.00  C148        C  
890      0      0      370376   7.75   NaN        Q  

 Columns are 

Index(['Survived', 'Pclass', 'Name', 'Sex', 'Age', 'Sibsp', 'Parch', 'Ticket',
       'Fare', 'Cabin', 'Embarked'],
      dtype='object')


Q1-Which features are categorical?

These values classify the samples into sets of similar samples. Within categorical features are the values nominal, ordinal, ratio, or interval based? Among other things this helps us select the appropriate plots for visualization.

Categorical: Survived, Sex, and Embarked. 
Ordinal: Pclass.

Q2-Which features are numerical? 
These values change from sample to sample. Within numerical features are the values discrete, continuous, or timeseries based? Among other things this helps us select the appropriate plots for visualization.

Continous: Age, Fare. 
Discrete: SibSp, Parch.

Q3-Which features are mixed data types?

Numerical, alphanumeric data within same feature. These are candidates for correcting goal.

Ticket is a mix of numeric and alphanumeric data types. 
Cabin is alphanumeric.



Q4- Which features may contain errors or typos?

This is harder to review for a large dataset, however reviewing a few samples from a smaller dataset may just tell us outright, which features may require correcting.

Name feature may contain errors or typos as there are several ways used to describe a name including titles, round brackets, and quotes used for alternative or short names.

Q5- Which features contain blank, null or empty values?

In [8]:
print(data_train.isnull().sum())

Survived      0
Pclass        0
Name          0
Sex           0
Age         177
Sibsp         0
Parch         0
Ticket        0
Fare          0
Cabin       686
Embarked      2
dtype: int64


These will require correcting.

Cabin > Age > Embarked features contain a number of null values in that order for the training dataset.
Cabin > Age are incomplete in case of test dataset.

Q6- What are the data types for various features?

In [11]:
data_train["Age"] = pd.to_numeric(data_train.Age, errors='coerce')
data_train["Age"].describe()

data_train["Fare"] = pd.to_numeric(data_train.Fare, errors='coerce')
data_train["Fare"].describe()

#data_train["Pid"] = pd.to_numeric(data_train.Pid, errors='coerce')
#data_train["Pid"].describe()

data_train["Survived"] = pd.to_numeric(data_train.Survived, errors='coerce')
data_train["Survived"].describe()


data_train["Pclass"] = pd.to_numeric(data_train.Pclass, errors='coerce')
data_train["Pclass"].describe()


data_train["Sibsp"] = pd.to_numeric(data_train.Sibsp, errors='coerce')
data_train["Sibsp"].describe()

data_train["Parch"] = pd.to_numeric(data_train.Parch, errors='coerce')
data_train["Parch"].describe()

##-----------------------------------------------------------------------##
##For Test Data set
##-----------------------------------------------------------------------##

data_test["Age"] = pd.to_numeric(data_test.Age, errors='coerce')
data_test["Age"].describe()

data_test["Fare"] = pd.to_numeric(data_test.Fare, errors='coerce')
data_test["Fare"].describe()

#data_test["Pid"] = pd.to_numeric(data_test.Pid, errors='coerce')
#data_test["Pid"].describe()

data_test["Pclass"] = pd.to_numeric(data_test.Pclass, errors='coerce')
data_test["Pclass"].describe()


data_test["Sibsp"] = pd.to_numeric(data_test.Sibsp, errors='coerce')
data_test["Sibsp"].describe()

data_test["Parch"] = pd.to_numeric(data_test.Parch, errors='coerce')
data_test["Parch"].describe()

data_train.info()
print('_'*40)
print('_'*40)
data_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 890 entries, 1 to 890
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  890 non-null    int64  
 1   Pclass    890 non-null    int64  
 2   Name      890 non-null    object 
 3   Sex       890 non-null    object 
 4   Age       713 non-null    float64
 5   Sibsp     890 non-null    int64  
 6   Parch     890 non-null    int64  
 7   Ticket    890 non-null    object 
 8   Fare      890 non-null    float64
 9   Cabin     204 non-null    object 
 10  Embarked  888 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 83.4+ KB
________________________________________
________________________________________
<class 'pandas.core.frame.DataFrame'>
Int64Index: 417 entries, 1 to 417
Data columns (total 10 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Pclass    417 non-null    int64  
 1   Name      417 no

Here are the results::
Seven features are floats. Six in case of test dataset.
Five features are strings (object).

Q7- What is the distribution of numerical feature values across the samples?

Catering null Values :::
Cabin & Fare is not very important feature for us
We will fill median values in age.

In [12]:
#data_train.isnull().sum() 
#Calculating Median
median = data_train['Age'].median()
data_train['Age'].fillna(median, inplace=True)
data_test['Age'].fillna(median, inplace=True)
data_train.isnull().sum() 



Survived      0
Pclass        0
Name          0
Sex           0
Age           0
Sibsp         0
Parch         0
Ticket        0
Fare          0
Cabin       686
Embarked      2
dtype: int64

This helps us determine, among other early insights, how representative is the training dataset of the actual problem domain.

1. Total samples are 891 or 40% of the actual number of passengers on board the Titanic (2,224). (agreed)
2. Survived is a categorical feature with 0 or 1 values. (agreed)
3. Around 38% samples survived representative of the actual survival rate at 32%.
4. Most passengers (> 75%) did not travel with parents or children.
5. Nearly 30% of the passengers had siblings and/or spouse aboard.
6. Fares varied significantly with few passengers (<1%) paying as high as $512.
7. Few elderly passengers (<1%) within age range 65-80.

In [13]:
data_train.describe()

Survived      Pclass         Age       Sibsp       Parch        Fare
count  890.000000  890.000000  890.000000  890.000000  890.000000  890.000000
mean     0.384270    2.307865   29.369854    0.522472    0.382022   32.232246
std      0.486696    0.836220   13.024675    1.103247    0.806409   49.714317
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   22.000000    0.000000    0.000000    7.925000
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   35.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [14]:
#Review survived rate using `percentiles=[.61, .62]` knowing our problem description mentions 38% survival rate.
percentile = [.61,.62] 
data_train["Survived"].describe(percentiles = percentile)

count    890.000000
mean       0.384270
std        0.486696
min        0.000000
50%        0.000000
61%        0.000000
62%        1.000000
max        1.000000
Name: Survived, dtype: float64

In [15]:
## Review Parch distribution using `percentiles=[.75, .8]`
percentile = [.75,.8] 
data_train["Parch"].describe(percentiles = percentile)


count    890.000000
mean       0.382022
std        0.806409
min        0.000000
50%        0.000000
75%        0.000000
80%        1.000000
max        6.000000
Name: Parch, dtype: float64

Q- What is the distribution of categorical features?


1. Names are unique across the dataset (count=unique=891)

2. Sex variable as two possible values with 64% male (top=male, freq=577/count=891).

3. Cabin values have several dupicates across samples. Alternatively several passengers shared a cabin.

4. Embarked takes three possible values. S port used by most passengers (top=S)

5. Ticket feature has high ratio (26%) of duplicate values (unique=681).

In [16]:
data_train.describe(include=['O'])

Name   Sex    Ticket        Cabin Embarked
count                             890   890       890          204      888
unique                            890     2       680          147        3
top     Skoog, Miss. Margit Elizabeth  male  CA. 2343  C23 C25 C27        S
freq                                1   576         7            4      643

Hint:::: 

DataFrame.describe(include=['O']) should include categorical columns only.

MOST IMPORTANT MARKDOWN ::

Asspumtions based on data analysis

We arrive at following assumptions based on data analysis done so far. We may validate these assumptions further before taking appropriate actions.

Correlating.

We want to know how well does each feature correlate with Survival. We want to do this early in our project and match these quick correlations with modelled correlations later in the project.

Completing.

We may want to complete Age feature as it is definitely correlated to survival.
We may want to complete the Embarked feature as it may also correlate with survival or another important feature.

Correcting.

Ticket feature may be dropped from our analysis as it contains high ratio of duplicates (22%) and there may not be a correlation between Ticket and survival.

Cabin feature may be dropped as it is highly incomplete or contains many null values both in training and test dataset.
PassengerId may be dropped from training dataset as it does not contribute to survival.

Name feature is relatively non-standard, may not contribute directly to survival, so maybe dropped.

Creating.

We may want to create a new feature called Family based on Parch and SibSp to get total count of family members on board.

We may want to engineer the Name feature to extract Title as a new feature.

We may want to create new feature for Age bands. This turns a continous numerical feature into an ordinal categorical feature.
(through Binning)

We may also want to create a Fare range feature if it helps our analysis.

Classifying.

We may also add to our assumptions based on the problem description noted earlier.

Women (Sex=female) were more likely to have survived.
Children (Age<?) were more likely to have survived.
The upper-class passengers (Pclass=1) were more likely to have survived.


So first of all we will analyze correlation between features:

1) Pclass with survival

In [ ]:
print(data_train.groupby('Pclass')['Pid'].count().sort_values(ascending=True).head())
data_train[['Pclass','Survived']].groupby(['Pclass'],as_index = False).mean().sort_values(by='Survived',ascending = False)


Pclass 3 has most no. of passengers (491) but they survived only 24%, 

However, Pclass 1 has 216 passengers, from which they survived most (62%).

2) Gender with survival

In [ ]:
print(data_train.groupby('Sex')['Pid'].count().sort_values(ascending=True).head())
data_train[['Sex','Survived']].groupby(['Sex'],as_index = False).mean().sort_values(by = 'Survived',ascending = False)

74% Females have survived.

3) sibsp with survival

In [ ]:
print(data_train.groupby('Sibsp')['Pid'].count().sort_values(ascending=True).head())
data_train[['Sibsp','Survived']].groupby(['Sibsp'],as_index =False).mean().sort_values(by = 'Survived',ascending = False)

53% with 1 Siblings have survived.
4) Parch with survival

In [ ]:
data_train[['Parch','Survived']].groupby(['Parch'],as_index = False).mean().sort_values(by = 'Survived',ascending = False)

60% with 3 arents/childrens have survived.


Analyze by visualizing data

Now we can continue confirming some of our assumptions using visualizations for analyzing the data.

Age is the most important feature in this dataset. So , lets visualize correlation between age and survival.

In [ ]:
# grid = sns.FacetGrid(train_df, col='Pclass', hue='Survived')
grid = sns.FacetGrid(data_train, col='Survived', size=5.2, aspect=1.6)
grid.map(plt.hist, 'Age', alpha=.5, bins=20)
grid.add_legend();

Correlating numerical features

Let us start by understanding correlations between numerical features and our solution goal (Survived).

Observations from above histogram are as follows:

1. Infants (Age <=4) had high survival rate.
2. Oldest passengers (Age = 80) survived.
3. Large number of 15-25 year olds did not survive.
4. Most passengers are in 15-35 age range.



Decisions.

This simple analysis confirms our assumptions as decisions for subsequent workflow stages.

1. We should consider Age in our model training.
2. Complete the Age feature for null values (Done already).
3. We should band age groups (Binning).

In [ ]:
# grid = sns.FacetGrid(train_df, col='Pclass', hue='Survived')
grid = sns.FacetGrid(data_train, col='Survived', row='Pclass', size=5.2, aspect=1.6)
grid.map(plt.hist, 'Age', alpha=.5, bins=20)
grid.add_legend();

Our Assumptions:
1) PClass 1 have survived most.
2) PClass 2 have greater infants survival.
3) PClass 3 have highest death rate, in addition, PClass 3 have most no. of passengers travelling.


Decisions.

Consider Pclass for model training.


In [ ]:
grid = sns.FacetGrid(data_train, row='Embarked', size=3.2, aspect=1.6)
grid.map(sns.pointplot, 'Pclass','Survived','Sex', palette='deep')
grid.add_legend();

Observations:
    1. In embarked 'S', female survial rate is high
    2. Exception in Embarked=C where males had higher survival rate. This could be a correlation between 
    Pclass and Embarked and in turn Pclass and Survived, not necessarily direct correlation between Embarked and Survived.
    3. Males had better survival rate in Pclass=3 when compared with Pclass=2 for C and Q ports.
    4. Ports of embarkation have varying survival rates for Pclass=3 and among male passengers.
    
Decisions.

    1. Add Sex feature to model training.
    2. Complete and add Embarked feature to model training.    

In [ ]:
grid = sns.FacetGrid(data_train, row='Embarked',col = 'Survived', size=3.2, aspect=1.6)
grid.map(sns.barplot, 'Sex','Fare', alpha=.5, ci=None)
grid.add_legend();

Observations.

1. Higher fare paying passengers had better survival.
2. Port of embarkation correlates with survival rates. 

Decisions.

Consider banding Fare feature.


## Wrangle data

We have collected several assumptions and decisions regarding our datasets and solution requirements. So far we did not have to change a single feature or value to arrive at these. Let us now execute our decisions and assumptions for correcting, creating, and completing goals.

# Correcting by dropping features

Based on our assumptions and decisions we want to drop the Cabin (correcting #2) and Ticket (correcting #1) features

Note that where applicable we perform operations on both training and testing datasets together to stay consistent.

In [17]:
combine = [data_train, data_test]
print("Before", data_train.shape, data_test.shape,combine[0].shape, combine[1].shape)

data_train = data_train.drop(['Ticket', 'Cabin'], axis=1)
data_test = data_test.drop(['Ticket', 'Cabin'], axis=1)
combine2 = [data_train, data_test]

print("After", data_train.shape, data_test.shape, combine2[0].shape, combine2[1].shape)

Before (890, 11) (417, 10) (890, 11) (417, 10)
After (890, 9) (417, 8) (890, 9) (417, 8)


First we will change name column, then apply binning to age column.

In [18]:
for dataset in combine2:
    dataset["Title"] = dataset.Name.str.extract(' ([A-Za-z]+)\.',expand=False)

In [19]:
data_train.head()

Survived  Pclass                                               Name  \
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2         1       3                             Heikkinen, Miss. Laina   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4         0       3                           Allen, Mr. William Henry   
5         0       3                                   Moran, Mr. James   

      Sex   Age  Sibsp  Parch     Fare Embarked Title  
1  female  38.0      1      0  71.2833        C   Mrs  
2  female  26.0      0      0   7.9250        S  Miss  
3  female  35.0      1      0  53.1000        S   Mrs  
4    male  35.0      0      0   8.0500        S    Mr  
5    male  28.0      0      0   8.4583        Q    Mr

In [20]:
pd.crosstab(data_train['Title'], data_train['Sex'])

Sex       female  male
Title                 
Capt           0     1
Col            0     2
Countess       1     0
Don            0     1
Dr             1     6
Jonkheer       0     1
Lady           1     0
Major          0     2
Master         0    40
Miss         182     0
Mlle           2     0
Mme            1     0
Mr             0   516
Mrs          125     0
Ms             1     0
Rev            0     6
Sir            0     1

We can replace many titles with a more common name or classify them as Rare.
For e.g. Mr to 1, Mrs to 2 so on.

In [21]:
for dataset in combine2:
    dataset['Title'] = dataset['Title'].replace(['Lady', 'Countess','Capt', 'Col','Don', 'Dr', 'Major', 'Rev', 'Sir', 'Jonkheer', 'Dona'], 'Rare')

    dataset['Title'] = dataset['Title'].replace('Mlle', 'Miss')
    dataset['Title'] = dataset['Title'].replace('Ms', 'Miss')
    dataset['Title'] = dataset['Title'].replace('Mme', 'Mrs')

In [22]:
data_train.head(10)

Survived  Pclass                                               Name  \
1          1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2          1       3                             Heikkinen, Miss. Laina   
3          1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4          0       3                           Allen, Mr. William Henry   
5          0       3                                   Moran, Mr. James   
6          0       1                            McCarthy, Mr. Timothy J   
7          0       3                     Palsson, Master. Gosta Leonard   
8          1       3  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)   
9          1       2                Nasser, Mrs. Nicholas (Adele Achem)   
10         1       3                    Sandstrom, Miss. Marguerite Rut   

       Sex   Age  Sibsp  Parch     Fare Embarked   Title  
1   female  38.0      1      0  71.2833        C     Mrs  
2   female  26.0      0      0   7.9250        S    Miss  
3   female  35.0      1      0  53.1000        S     Mrs  
4     male  35.0      0      0   8.0500        S      Mr  
5     male  28.0      0      0   8.4583        Q      Mr  
6     male  54.0      0      0  51.8625        S      Mr  
7     male   2.0      3      1  21.0750        S  Master  
8   female  27.0      0      2  11.1333        S     Mrs  
9   female  14.0      1      0  30.0708        C     Mrs  
10  female   4.0      1      1  16.7000        S    Miss

In [23]:
data_train[['Title','Survived']].groupby(['Title'],as_index = False).mean()

Title  Survived
0  Master  0.575000
1    Miss  0.702703
2      Mr  0.156977
3     Mrs  0.793651
4    Rare  0.347826

In [24]:
#We can convert the categorical titles to ordinal.

titlemapping = {'Mr':1, 'Miss':2,'Mrs':3, 'Master':4,'Rare':5}
for row in combine2:
    row["Title"] = row["Title"].map(titlemapping)
    row['Title'] = row['Title'].fillna(0)
    

In [25]:
data_train.head(10)

Survived  Pclass                                               Name  \
1          1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2          1       3                             Heikkinen, Miss. Laina   
3          1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4          0       3                           Allen, Mr. William Henry   
5          0       3                                   Moran, Mr. James   
6          0       1                            McCarthy, Mr. Timothy J   
7          0       3                     Palsson, Master. Gosta Leonard   
8          1       3  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)   
9          1       2                Nasser, Mrs. Nicholas (Adele Achem)   
10         1       3                    Sandstrom, Miss. Marguerite Rut   

       Sex   Age  Sibsp  Parch     Fare Embarked  Title  
1   female  38.0      1      0  71.2833        C      3  
2   female  26.0      0      0   7.9250        S      2  
3   female  35.0      1      0  53.1000        S      3  
4     male  35.0      0      0   8.0500        S      1  
5     male  28.0      0      0   8.4583        Q      1  
6     male  54.0      0      0  51.8625        S      1  
7     male   2.0      3      1  21.0750        S      4  
8   female  27.0      0      2  11.1333        S      3  
9   female  14.0      1      0  30.0708        C      3  
10  female   4.0      1      1  16.7000        S      2

In [27]:
#Now we can drop Name column. We also do not need the PassengerId feature in the training dataset.

data_train = data_train.drop(['Name'],axis = 1)
data_test = data_test.drop(['Name'],axis = 1)
combine2 = [data_train, data_test]
data_train.shape, data_test.shape
#data_train.head(10)

((890, 9), (417, 8))

In [28]:
#Lets change Sex Column
titlemapping = {'male':0, 'female':1}
for row in combine2:
    row["Sex"] = row["Sex"].map(titlemapping).astype(int)
    
    row['Sex'] = row['Sex'].fillna(0)

In [29]:
data_train.head(10)

Survived  Pclass  Sex   Age  Sibsp  Parch     Fare Embarked  Title
1          1       1    1  38.0      1      0  71.2833        C      3
2          1       3    1  26.0      0      0   7.9250        S      2
3          1       1    1  35.0      1      0  53.1000        S      3
4          0       3    0  35.0      0      0   8.0500        S      1
5          0       3    0  28.0      0      0   8.4583        Q      1
6          0       1    0  54.0      0      0  51.8625        S      1
7          0       3    0   2.0      3      1  21.0750        S      4
8          1       3    1  27.0      0      2  11.1333        S      3
9          1       2    1  14.0      1      0  30.0708        C      3
10         1       3    1   4.0      1      1  16.7000        S      2

In [30]:
data_train.to_csv("training-data.csv")
data_test.to_csv("testing-data.csv")

After these All EDA will work on our project.............!!!!!!

In [ ]:
#Binning of age

data_train["Ageband"] = pd.cut(data_train["Age"],5)
data_train[['Ageband','Survived']].groupby(['Ageband'],as_index = False).mean().sort_values(by='Ageband', ascending=True)

In [ ]:
#Let us replace Age with ordinals based on these bands.

for row in combine2:
    row.loc[row["Age"] <= 16.336,'Age'] = 0
    row.loc[(row['Age'] > 16.336) & (row["Age"] <= 32.252),'Age'] = 1
    
    row.loc[(row["Age"] > 32.252) & (row["Age"] <= 48.168),'Age'] = 2
    row.loc[(row["Age"] > 48.168) & (row["Age"] <= 64.084),'Age'] = 3
    row.loc[row["Age"] > 64.084,'Age' ] = 4


In [ ]:
data_train["Age"] = data_train["Age"].astype(int)
data_train.head(10)

In [ ]:
#Delete Age band column from tarin dataset
data_train = data_train.drop(['Ageband'],axis = 1)
#data_test = data_test.drop(['Name'],axis = 1)
combine2 = [data_train, data_test]
data_train.shape, data_test.shape

Create new feature combining existing features
We can create a new feature for "FamilySize" which combines Parch and SibSp. 

This will enable us to drop Parch and SibSp from our datasets.

In [ ]:
#for row in combine2:
    #row['FamilySize'] = row['Sibsp'] + row['Parch']+1
for dataset in combine2:
    
    dataset["FamilySize"] = dataset['Sibsp'] + dataset['Parch'] + 1

In [ ]:
data_train.head()


In [ ]:
data_train[["FamilySize","Survived"]].groupby(["FamilySize"],as_index= False).mean().sort_values(by='Survived', ascending=False)

In [ ]:
#We can create another feature called IsAlone.
#iS ALONE MEANS =>>>> IF HE HAS FAMILY MEMBER EQUAL TO 1 THEN ISALONE VALUE IS 1 (FALSE) ELSE 0 (iSALONE TRUE)
#0 means yes , 1 means no
#isalon = 0 means yes, he is alone
for row in combine2:
    row["IsAlone"] = 0
    row.loc[row["FamilySize"] == 1,'IsAlone'] = 1
    

In [ ]:
data_train[['IsAlone','Survived']].groupby(['IsAlone'],as_index = False).mean()

In [ ]:
#Let us drop Parch, SibSp, and FamilySize features in favor of IsAlone.
data_train = data_train.drop(['Parch', 'Sibsp', 'FamilySize'], axis=1)
data_test = data_test.drop(['Parch', 'Sibsp', 'FamilySize'], axis=1)
combine2 = [data_train, data_test]

In [ ]:
data_train.head()

In [ ]:
#We can also create an artificial feature combining Pclass and Age.
for row in combine2:
    row['Age*Class'] = row.Age * row.Pclass


Take median for null values in embarked

In [ ]:
freq_port = data_train.Embarked.dropna().mode()[0]
freq_port

In [ ]:
#Replace S as median
for dataset in combine2:
    dataset['Embarked'] = dataset['Embarked'].fillna(freq_port)

Converting categorical feature to numeric

We can now convert the EmbarkedFill feature by creating a new numeric Port feature.

In [ ]:
titlemapping = {'S':0, 'C':1,'Q':3}
for row in combine2:
    row["Embarked"] = row["Embarked"].map(titlemapping).astype(int)
    

In [ ]:
data_train.head()

In [ ]:


AvgFare = data_train['Fare'].median()
data_train['Fare'].fillna(AvgFare, inplace=True)
data_test['Fare'].fillna(AvgFare, inplace=True)

data_train.head()


Create FareBand, first check its cut values

In [ ]:
data_train['FareBand'] = pd.qcut(data_train['Fare'], 4)
data_train[['FareBand','Survived']].groupby(['FareBand'],as_index = False).mean()

In [ ]:
#Now Create bining inthe given format above
for row in combine2:
    row.loc[row['Fare'] <= 7.91,'Fare'] = 0
    row.loc[(row['Fare']>7.91) & (row['Fare']<= 14.54),'Fare'] = 1
    row.loc[(row['Fare']>14.54) & (row['Fare']<= 31.0),'Fare'] = 2
    row.loc[(row['Fare']>31.0) & (row['Fare']<= 512.329),'Fare'] = 3
data_train['Fare'] = data_train['Fare'].astype(int)

In [ ]:
data_train = data_train.drop(['FareBand'], axis=1)
#data_test = data_test.drop(['Parch', 'Sibsp', 'FamilySize'], axis=1)
combine2 = [data_train, data_test]

In [ ]:
#Preprocessed and prepared training data set for model prediction is here:::::
data_train.head()

In [ ]:

#Preprocessed and prepared TEST data set for model prediction is here:::::
data_test.head()

Model, predict and solve


Now we are ready to train a model and predict the required solution. There are 60+ predictive modelling algorithms to choose from. We must understand the type of problem and solution requirement to narrow down to a select few models which we can evaluate. Our problem is a classification and regression problem. We want to identify relationship between output (Survived or not) with other variables or features (Gender, Age, Port...). We are also perfoming a category of machine learning which is called supervised learning as we are training our model with a given dataset. With these two criteria - Supervised Learning plus Classification and Regression, we can narrow down our choice of models to a few. These include:

Logistic Regression
KNN or k-Nearest Neighbors
Support Vector Machines
Naive Bayes classifier
Decision Tree
Random Forrest
Perceptron
Artificial neural network
RVM or Relevance Vector Machine

Firstly Checking null values in test data and train data set:

In [ ]:
print(data_test.isnull().sum())

In [ ]:
print(data_train.isnull().sum())

In [ ]:
X_Train = data_train.drop("Survived",axis =1)
Y_Train = data_train["Survived"]
X_Test = data_test.drop(["Pid"],axis =1)

X_Train.shape,Y_Train.shape,X_Test.shape

In [ ]:
# Logistic Regression

logreg = LogisticRegression()
logreg.fit(X_Train, Y_Train)
Y_pred = logreg.predict(X_Test)
acc_log = round(logreg.score(X_Train, Y_Train) * 100, 2)
acc_log

Next we model using Support Vector Machines which are supervised learning models with associated learning algorithms that analyze data used for classification and regression analysis. Given a set of training samples, each marked as belonging to one or the other of two categories, an SVM training algorithm builds a model that assigns new test samples to one category or the other, making it a non-probabilistic binary linear classifier. Reference Wikipedia.

Note that the model generates a confidence score which is higher than Logistics Regression model.

In [ ]:
# Support Vector Machines

svc = SVC()
svc.fit(X_Train, Y_Train)
Y_pred = svc.predict(X_Test)
acc_svc = round(svc.score(X_Train, Y_Train) * 100, 2)
acc_svc

In pattern recognition, the k-Nearest Neighbors algorithm (or k-NN for short) is a non-parametric method used for classification and regression. A sample is classified by a majority vote of its neighbors, with the sample being assigned to the class most common among its k nearest neighbors (k is a positive integer, typically small). If k = 1, then the object is simply assigned to the class of that single nearest neighbor. Reference Wikipedia.

KNN confidence score is better than Logistics Regression but worse than SVM.



In [ ]:
knn = KNeighborsClassifier(n_neighbors = 3)
knn.fit(X_Train, Y_Train)
Y_pred = knn.predict(X_Test)
acc_knn = round(knn.score(X_Train, Y_Train) * 100, 2)
print(Y_pred)
acc_knn


In machine learning, naive Bayes classifiers are a family of simple probabilistic classifiers based on applying Bayes' theorem 
with strong (naive) independence assumptions between the features. Naive Bayes classifiers are highly scalable, 
requiring a number of parameters linear in the number of variables (features) in a learning problem. 

The model generated confidence score is the lowest among the models evaluated so far.

In [ ]:
# Gaussian Naive Bayes
gaussian = GaussianNB()
gaussian.fit(X_Train, Y_Train)
Y_pred = gaussian.predict(X_Test)
acc_gaussian = round(gaussian.score(X_Train, Y_Train) * 100, 2)
acc_gaussian

The perceptron is an algorithm for supervised learning of binary classifiers (functions that can decide whether an input, represented by a vector of numbers, belongs to some specific class or not). It is a type of linear classifier, i.e. a classification algorithm that makes its predictions based on a linear predictor function combining a set of weights with the feature vector. The algorithm allows for online learning, in that it processes elements in the training set one at a time. Reference Wikipedia.

In [ ]:
# Perceptron
perceptron = Perceptron()
perceptron.fit(X_Train, Y_Train)
Y_pred = perceptron.predict(X_Test)
acc_Percep = round(perceptron.score(X_Train, Y_Train) * 100, 2)
acc_Percep

In [ ]:
# Linear SVC
Linear_svc = LinearSVC()
Linear_svc.fit(X_Train,Y_Train)
Y_pred = Linear_svc.predict(X_Test)
accLinearSVC = round(Linear_svc.score(X_Train,Y_Train)*100,2)
accLinearSVC

In [ ]:
# Stochastic Gradient Descent
sgd = SGDClassifier()
sgd.fit(X_Train, Y_Train)
Y_pred = sgd.predict(X_Test)
acc_sgd = round(sgd.score(X_Train, Y_Train) * 100, 2)
acc_sgd


This model uses a decision tree as a predictive model which maps features (tree branches) to conclusions about the target value (tree leaves). Tree models where the target variable can take a finite set of values are called classification trees; in these tree structures, leaves represent class labels and branches represent conjunctions of features that lead to those class labels. Decision trees where the target variable can take continuous values (typically real numbers) are called regression trees. Reference Wikipedia.

The model confidence score is the highest among models evaluated so far.

In [ ]:
# Decision Tree

decision_tree = DecisionTreeClassifier()
decision_tree.fit(X_Train, Y_Train)
Y_pred = decision_tree.predict(X_Test)
acc_decision_tree = round(decision_tree.score(X_Train, Y_Train) * 100, 2)
acc_decision_tree

The next model Random Forests is one of the most popular. Random forests or random decision forests are an ensemble learning method for classification, regression and other tasks, that operate by constructing a multitude of decision trees (n_estimators=100) at training time and outputting the class that is the mode of the classes (classification) or mean prediction (regression) of the individual trees. Reference Wikipedia.

The model confidence score is the highest among models evaluated so far. We decide to use this model's output (Y_pred) for creating our competition submission of results.

In [ ]:
# Random Forest

random_forest = RandomForestClassifier(n_estimators=100)
random_forest.fit(X_Train, Y_Train)
Y_pred = random_forest.predict(X_Test)
random_forest.score(X_Train, Y_Train)
acc_random_forest = round(random_forest.score(X_Train, Y_Train) * 100, 2)
acc_random_forest

Model evaluation

We can now rank our evaluation of all the models to choose the best one for our problem. While both Decision Tree and Random Forest score the same, we choose to use Random Forest as they correct for decision trees' habit of overfitting to their training set.

In [ ]:
models = pd.DataFrame({
    'Model': ['Support Vector Machines', 'KNN', 'Logistic Regression', 
              'Random Forest', 'Naive Bayes', 'Perceptron', 
              'Stochastic Gradient Decent', 'Linear SVC', 
              'Decision Tree'],
    'Score': [acc_svc, acc_knn, acc_log, 
              acc_random_forest, acc_gaussian, acc_Percep, 
              acc_sgd, acc_linear_svc, acc_decision_tree]})
models.sort_values(by='Score', ascending=False)

In [ ]:
submission = pd.DataFrame({
        "PassengerId": data_test["Pid"],
        "Survived": Y_pred
    })
submission.to_csv('Titanic-project-kaggle/submission_Kaggle.csv', index=False)